src/services/TrainedQuestionGenerator.py

```diff
@@ -1,3 +1,90 @@
+from src.models import db
+from src.models.GeneratedQuestionModel import GeneratedQuestionModel
+from src.models.IdiomModel import IdiomModel
+from src.models.UserAnswerModel import UserAnswerModel
+
+import random
+
+
+def generate_question(user_id):
+    answered_questions = get_user_answered_questions(user_id)
+
+    if len(answered_questions) < 1:
+        return generate_first_question()
+    else:
+        return generate_question_based_on_success_rate(answered_questions)
+
+
+def get_sorted_idioms_with_frequency():
+    return IdiomModel.query.order_by(IdiomModel.frequency.desc()).all()
+
+
+def get_user_answered_questions(user_id):
+    answered_questions = (db.session.query(UserAnswerModel, GeneratedQuestionModel).filter(
+        UserAnswerModel.user_id == user_id).filter(
+        GeneratedQuestionModel.id == UserAnswerModel.generated_question_id)
+        .order_by(GeneratedQuestionModel.created_at.desc())
+        .all())
+
+    return answered_questions
+
+
+def generate_first_question():
+    sorted_idioms = get_sorted_idioms_with_frequency()
+
+    return sorted_idioms[int(round(len(sorted_idioms)/2))]
+
+
+def generate_question_based_on_success_rate(answered_questions):
+    sorted_idioms = list(get_sorted_idioms_with_frequency())
+
+    last_answered_question = answered_questions[0]
+    position = get_position_of_last_question(sorted_idioms, last_answered_question)
+
+    rate = success_rate(answered_questions)
+    if rate > 0.9:
+        return sorted_idioms[get_new_question_position(position + 61)]
+    elif rate > 0.8:
+        return sorted_idioms[get_new_question_position(position + 21)]
+    elif rate > 0.6:
+        return sorted_idioms[get_new_question_position(position - 21)]
+    elif rate > 0.4:
+        return sorted_idioms[get_new_question_position(position - 31)]
+    else:
+        return sorted_idioms[get_new_question_position(position - 61)]
```

Comment:
Trochu sa bojim hranicnych hodnot.. ak uzivatel nespravne odpovie na najlahsiu otazku, tak ho posunies o 61 do minusu, kedze osetrujes iba spodnu hranicu, tak vrchna ostane -51... pri minime tam mozes prihodit abs a budes to mat celkom vyriesene, pri maxime si skontroluj spodny aj vrchny bound na to, ci su mensie ako MAX_IDIOM_ID a ak nie su, tak im ten range zvacsi na priklad aj na vsetky idiomy s pocetnostou 0

Reply:
Ked tak urob test s uzivatelom, co odpoveda vzdy nespravne a vzdy spravne, aby si zistila, ci sa ti to niekde na tych hranicnych nezadrhne (spravne odpovede vyhladas v jsone so vsetkymi odpovedami)